extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java

```diff
@@ -757,34 +861,39 @@ private void checkTaskDuration()
       // if this task has run longer than the configured duration, signal all tasks in the group to persist
       if (earliestTaskStart.plus(ioConfig.getTaskDuration()).isBeforeNow()) {
         log.info("Task group [%d] has run for [%s]", groupId, ioConfig.getTaskDuration());
-        Map<Integer, Long> endOffsets = signalTasksToFinish(groupId);
+        futureGroupIds.add(groupId);
+        futures.add(signalTasksToFinish(groupId));
+      }
+    }
 
+    List<Map<Integer, Long>> results = Futures.successfulAsList(futures).get();
+    for (int j = 0; j < results.size(); j++) {
+      Integer groupId = futureGroupIds.get(j);
+      TaskGroup group = taskGroups.get(groupId);
+      Map<Integer, Long> endOffsets = results.get(j);
+
+      if (endOffsets != null) {
         // set a timeout and put this group in pendingCompletionTaskGroups so that it can be monitored for completion
         group.completionTimeout = DateTime.now().plus(ioConfig.getCompletionTimeout());
-        if (!pendingCompletionTaskGroups.containsKey(groupId)) {
-          pendingCompletionTaskGroups.put(groupId, Lists.<TaskGroup>newArrayList());
-        }
+        pendingCompletionTaskGroups.putIfAbsent(groupId, Lists.<TaskGroup>newCopyOnWriteArrayList());
         pendingCompletionTaskGroups.get(groupId).add(group);
 
-        // if we know what the endOffsets are going to be from talking to the tasks, set them as the next startOffsets
-        if (endOffsets != null) {
-          for (Map.Entry<Integer, Long> entry : endOffsets.entrySet()) {
-            partitionGroups.get(groupId).put(entry.getKey(), entry.getValue());
-          }
+        // set endOffsets as the next startOffsets
+        for (Map.Entry<Integer, Long> entry : endOffsets.entrySet()) {
+          partitionGroups.get(groupId).put(entry.getKey(), entry.getValue());
         }
-
-        // remove this task group from the list of current task groups now that it has been handled
-        i.remove();
       }
+
+      // remove this task group from the list of current task groups now that it has been handled
```

Comment:
we probably need to handle the case when endOffsets is null, i.e the future threw exception. 
wonder if we should hard kill the task in that case ?